src/particles/ShapeFactors.H

```diff
@@ -107,83 +107,52 @@ int compute_shape_factor <3> (amrex::Real* const sx, amrex::Real xmid){
- *  Specialization for order 1
- */
-template <>
-AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-int compute_shifted_shape_factor <1> (amrex::Real* const sx,
-                                      const amrex::Real x_old,
-                                      const int i_new){
-
-    using namespace amrex::literals;
+struct shape_factor_result {
+    amrex::Real factor;
+    int cell;
+};
 
-    const auto i = static_cast<int>(amrex::Math::floor(x_old));
-    const int i_shift = i - i_new;
-    const amrex::Real xint = x_old - i;
-    sx[1+i_shift] = 1.0_rt - xint;
-    sx[2+i_shift] = xint;
-    return i;
-}
-
-/**
- *  Compute shifted shape factor and return index of leftmost cell where
- *  particle writes, for Esirkepov algorithm.
- *  Specialization for order 2
- */
-template <>
-AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-int compute_shifted_shape_factor <2> (amrex::Real* const sx,
-                                      const amrex::Real x_old,
-                                      const int i_new){
-
-    using namespace amrex::literals;
-
-    const auto i = static_cast<int>(amrex::Math::floor(x_old+0.5_rt));
-    const int i_shift = i - (i_new + 1);
-    const amrex::Real xint = x_old - i;
-    sx[1+i_shift] = 0.5_rt*(0.5_rt-xint)*(0.5_rt-xint);
-    sx[2+i_shift] = 0.75_rt-xint*xint;
-    sx[3+i_shift] = 0.5_rt*(0.5_rt+xint)*(0.5_rt+xint);
-    // index of the leftmost cell where particle deposits
-    return i-1;
-}
-
-/**
- *  Compute shifted shape factor and return index of leftmost cell where
- *  particle writes, for Esirkepov algorithm.
- *  Specialization for order 3
+/** \brief Compute a single shape factor and return the index of the cell where the particle writes.
+ *
+ * \tparam depos_order Order of the shape factor
+ * \param[in] xmid exact position of the particle in index space
+ * \param[in] ix index of the shape factor, must be 0 <= ix <= depos_order
  */
-template <>
-AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-int compute_shifted_shape_factor <3> (amrex::Real* const sx,
-                                      const amrex::Real x_old,
-                                      const int i_new){
+template<int depos_order> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+shape_factor_result compute_shape_factor_branchless (amrex::Real xmid, int ix) noexcept {
 
     using namespace amrex::literals;
 
-    const auto i = static_cast<int>(amrex::Math::floor(x_old));
-    const int i_shift = i - (i_new + 1);
-    const amrex::Real xint = x_old - static_cast<amrex::Real>(i);
-    sx[1+i_shift] = 1.0_rt/6.0_rt*(1.0_rt-xint)*(1.0_rt-xint)*(1.0_rt-xint);
-    sx[2+i_shift] = 2.0_rt/3.0_rt-xint*xint*(1.0_rt-xint/2.0_rt);
-    sx[3+i_shift] = 2.0_rt/3.0_rt-(1.0_rt-xint)*(1.0_rt-xint)*(1.0_rt-0.5_rt*(1.0_rt-xint));
-    sx[4+i_shift] = 1.0_rt/6.0_rt*xint*xint*xint;
-    // index of the leftmost cell where particle deposits
-    return i-1;
+    if constexpr (depos_order==0) {
+        const int i = static_cast<int>(std::floor(xmid+0.5));
```

Comment:
to avoid implicit conversions from `double` to `float` when running SP.
```suggestion
        const int i = static_cast<int>(std::floor(xmid+0.5_rt));
```